<a href="https://colab.research.google.com/github/WaleedSharkas/hackerfile/blob/master/Exploring_Ebay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving autos.csv to autos (3).csv


# Introduction 

In this project, we will work with a dataset on used cars from eBay Kleinanzeigen, a classifieds section from the Germany eBay website. Although the full dataset consisted of over 370,000 listings, we will only use 50,000 data points. This will allow us to demonstrate how to effectively clean a data set from scratch.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken 
from this date.
-name - Name of the car.
-seller - Whether the seller is private or a dealer.
-offerType - The type of listing
-price - The price on the ad to sell the car.
-abtest - Whether the listing is included in an A/B test.
-vehicleType - The vehicle Type.
-yearOfRegistration - The year in which which year the car was first registered.
-gearbox - The transmission type.
-powerPS - The power of the car in PS.
-model - The car model name.
-kilometer - How many kilometers the car has driven.
-monthOfRegistration - The month in which which year the car was first registered.
-fuelType - What type of fuel the car uses.
-brand - The brand of the car.
-notRepairedDamage - If the car has a damage which is not yet repaired.
-dateCreated - The date on which the eBay listing was created.
-nrOfPictures - The number of pictures in the ad.
-postalCode - The postal code for the location of the vehicle.
-lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean the data and then use pandas to perform some basic initial analysis of the listings.

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

Observations on current data:
- Contains 20 columns, most stored as strings
- Some Columns contain null values
- dateCrawled, LastSeen, price, odometer are stored as strings. Will need to be changed 
- certain coloumns will need to be re-named and converted from camelcase to snakecase


In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns= ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

We cleaned up the columns to make them less confusing and more clean:

-yearOfRegistration to registration_year

-monthOfRegistration to registration_month

-notRepairedDamage to unrepaired_damage

-dateCreated to ad_created

Also changed the columns from camelcase to snakecase due to it being the preferred style for Python.

In [6]:
autos.describe(include='all')

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-14 20:50:02  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]

In [7]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [9]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [10]:
pd.to_datetime(autos["date_crawled"]).max()
# most recent date an observation was collected.

Timestamp('2016-04-07 14:36:56')

In [11]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].count()
# number of cars with registration years prior to 1906 or after 2016

1972

In [12]:
 # the number of cars with power rating less than 1ps or greater than 2000ps
autos.loc[~autos["power_ps"].between(1, 2000), "power_ps"].count()

5518

In [13]:
autos["registration_month"].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

Observations:
- we notice that the seller and offer_type columns are almost exclusively a single value. That means they are good candidates to drop as they won't have useful information for analysis. 
-It also appears that the column nr_of_pictures is all zeros, since the minimum and maximum values in the column are both zero, good candidates to drop as they won't have useful information for analysis. 
- the power_ps column has suspicious amount below 1ps
- spicious minimum value of 0, due to There being no zero month. Could be due to typos

In [14]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [15]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [17]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [18]:
print(autos["odometer_km"].nunique())

13


In [19]:
(autos["odometer_km"] >= 100000).sum()

39763

Overall, there doesnt seem to be anything that stands out in the odometer_km data.

In [20]:
autos["price"].nunique()

2357

In [21]:
print(autos["price"].describe())
autos["price"].value_counts().head(15)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

In [22]:
autos["price"].value_counts().sort_index(ascending=True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64

In [23]:
autos["price"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

The price column has numbers in both the max and min list that do not seem unrealistic. Further breakdown of numbers is needed here.

In [24]:
autos.loc[autos["price"] > 350000, ["name", "model", "brand", "registration_year", "price"]]

name  ...     price
514                     Ford_Focus_Turnier_1.6_16V_Style  ...    999999
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  ...  11111111
7814                                         Ferrari_F40  ...   1300000
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  ...  10000000
22947                         Bmw_530d_zum_ausschlachten  ...   1234566
24384                            Schlachte_Golf_3_gt_tdi  ...  11111111
27371                                         Fiat_Punto  ...  12345678
37585                                Volkswagen_Jetta_GT  ...    999990
39377                        Tausche_volvo_v40_gegen_van  ...  12345678
39705                        Tausch_gegen_gleichwertiges  ...  99999999
42221                                  Leasinguebernahme  ...  27322222
43049                                      2_VW_Busse_T3  ...    999999
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  ...  12345678
47634                                        Ferrari_FXX  ...   3890000

[14 rows x 5 columns]

In [25]:
autos.loc[autos["price"] < 0, ["name", "model", "brand", "registration_year", "price"]]

Empty DataFrame
Columns: [name, model, brand, registration_year, price]
Index: []

Price Observations:
- There are 1,421 cars listed with $0 price, It is only 2% of data. We will wait to see if it needs to be removed. 
- It does seem odd that a ford focus would sell for that high. Not sure if it is worth confirming this information. 

In [26]:
autos[["date_crawled", "ad_created", "last_seen"]] = autos[["date_crawled", "ad_created", "last_seen"]].apply(pd.to_datetime)

In [27]:
autos[["date_crawled", "ad_created", "last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date_crawled  50000 non-null  datetime64[ns]
 1   ad_created    50000 non-null  datetime64[ns]
 2   last_seen     50000 non-null  datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 1.1 MB


In [28]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled ad_created           last_seen
0 2016-03-26 17:47:46 2016-03-26 2016-04-06 06:45:54
1 2016-04-04 13:38:56 2016-04-04 2016-04-06 14:45:08
2 2016-03-26 18:57:24 2016-03-26 2016-04-06 20:15:37
3 2016-03-12 16:58:10 2016-03-12 2016-03-15 03:16:28
4 2016-04-01 14:38:50 2016-04-01 2016-04-01 14:38:50

In [29]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [30]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_values()

2016-04-07    0.00142
2016-04-06    0.00318
2016-03-18    0.01306
2016-04-05    0.01310
2016-03-06    0.01394
2016-03-13    0.01556
2016-03-05    0.02538
2016-03-24    0.02910
2016-03-16    0.02950
2016-03-27    0.03104
2016-03-17    0.03152
2016-03-25    0.03174
2016-03-31    0.03192
2016-03-10    0.03212
2016-03-23    0.03238
2016-03-26    0.03248
2016-03-11    0.03248
2016-03-22    0.03294
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-30    0.03362
2016-04-01    0.03380
2016-03-15    0.03398
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-19    0.03490
2016-04-02    0.03540
2016-03-07    0.03596
2016-04-04    0.03652
2016-03-14    0.03662
2016-03-12    0.03678
2016-03-21    0.03752
2016-03-20    0.03782
2016-04-03    0.03868
Name: date_crawled, dtype: float64

In [31]:
autos["date_crawled"].dt.hour.value_counts(normalize = True).sort_index()

0     0.02716
1     0.01542
2     0.00652
3     0.00502
4     0.00420
5     0.00310
6     0.00350
7     0.01442
8     0.01994
9     0.03556
10    0.04812
11    0.05540
12    0.05884
13    0.06466
14    0.06808
15    0.06626
16    0.06960
17    0.06988
18    0.07356
19    0.06792
20    0.06780
21    0.06148
22    0.05088
23    0.04268
Name: date_crawled, dtype: float64

Data Crawled Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. The overall distribution is faily uniformly scrpaed over. Hourly breakdown shows non uniformly between 2am-6am. 

In [32]:
(autos["last_seen"].dt.floor("D").value_counts(normalize=True).sort_index())

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

Last Seen Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. Over half of the entries were last seen in the last five days of data.

In [33]:
autos["ad_created"].dt.floor("D").value_counts(normalize = True).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [34]:
autos["ad_created"].dt.year.value_counts(normalize = True)

2016    0.99988
2015    0.00012
Name: ad_created, dtype: float64

In [35]:
autos["ad_created"].dt.hour.value_counts(normalize = True).sort_index()

0    1.0
Name: ad_created, dtype: float64

AD Created Observation
 
The ads were created between june 2015  and april 2016. 

In [36]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registratoin Year Observations

There are odd numbers here, it is not possible to have a car in the "1000" which is the min value. The max value is also odd with "9999". Any vehivle registred after 2016 should not be in the data. We should focus on 1900-2016. Further Breakdown needed.

In [37]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: registration_year, dtype: int64

In [38]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]
# Percent of invalid data 

0.03944

In [39]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

In [40]:
autos.loc[autos["registration_year"].between(1995, 2016), "registration_year"].value_counts(normalize = True).sort_index()

1995    0.029579
1996    0.032531
1997    0.045687
1998    0.055261
1999    0.067584
2000    0.075559
2001    0.060893
2002    0.057064
2003    0.061434
2004    0.061659
2005    0.067922
2006    0.061006
2007    0.051905
2008    0.050260
2009    0.047264
2010    0.035977
2011    0.036811
2012    0.029805
2013    0.018158
2014    0.015004
2015    0.008989
2016    0.029647
Name: registration_year, dtype: float64

In [41]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].value_counts(normalize = True).sort_index().loc[1970:].sum()

0.9952319480303154

Registratoin Year Observations Final

Due to around 99% of this data occuring after 1970. We will set a date range for this data " 1970 - 2016".

In [42]:
autos = autos[autos["registration_year"].between(1969,2016)]
autos["registration_year"].value_counts(normalize=True).head(30)

2000    0.070141
2005    0.063052
1999    0.062738
2004    0.057238
2003    0.057029
2006    0.056631
2001    0.056527
2002    0.052972
1998    0.051299
2007    0.048183
2008    0.046656
2009    0.043875
1997    0.042411
2011    0.034171
2010    0.033397
1996    0.030198
2012    0.027667
2016    0.027521
1995    0.027458
2013    0.016856
2014    0.013928
1994    0.013802
1993    0.009306
2015    0.008344
1990    0.008260
1992    0.008177
1991    0.007445
1989    0.003785
1988    0.002970
1985    0.002196
Name: registration_year, dtype: float64

In [43]:
autos["brand"].nunique()


40

In [44]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212723
bmw               0.110377
opel              0.108432
mercedes_benz     0.095278
audi              0.086746
ford              0.069221
renault           0.047493
peugeot           0.029633
fiat              0.025827
seat              0.018236
skoda             0.016061
mazda             0.015203
nissan            0.015162
smart             0.013970
citroen           0.013949
toyota            0.012527
sonstige_autos    0.009933
hyundai           0.009892
volvo             0.009139
mini              0.008679
mitsubishi        0.008177
honda             0.007884
kia               0.007131
alfa_romeo        0.006608
porsche           0.005939
suzuki            0.005939
chevrolet         0.005542
chrysler          0.003639
dacia             0.002572
daihatsu          0.002572
jeep              0.002259
subaru            0.002196
land_rover        0.002008
saab              0.001610
jaguar            0.001548
daewoo            0.001506
rover             0.001359
t

In [45]:
autos["brand"].value_counts(normalize = True).head(10).sum()

0.8039650340875821

Brand Observations
- Volkswagen is the most popular brand, with aprox double bmw who is second most popular
- WE will limit our analysis to brands in the top 10 ">0.018"

In [46]:
brand_count = autos["brand"].value_counts(normalize=True)
common_brands = brand_count[brand_count > .018].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [47]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9088,
 'bmw': 8338,
 'fiat': 2680,
 'ford': 7073,
 'mercedes_benz': 30330,
 'opel': 5254,
 'peugeot': 3033,
 'renault': 2390,
 'seat': 4301,
 'volkswagen': 6508}

In [48]:
autos.groupby("brand")["model"].value_counts(dropna = False).sort_values(ascending = False).head(10)

brand          model   
volkswagen     golf        3815
bmw            3er         2688
volkswagen     polo        1677
opel           corsa       1644
               astra       1388
volkswagen     passat      1388
audi           a4          1265
bmw            5er         1163
mercedes_benz  c_klasse    1147
               e_klasse     981
Name: model, dtype: int64

In [49]:
autos.groupby("odometer_km")["brand"].describe().sort_index().round(1)

count unique         top  freq
odometer_km                                
5000           872     39  volkswagen   180
10000          244     26  volkswagen    48
20000          754     36  volkswagen   141
30000          763     35  volkswagen   168
40000          795     34  volkswagen   143
50000          997     38  volkswagen   184
60000         1122     39  volkswagen   206
70000         1186     39  volkswagen   208
80000         1386     38  volkswagen   252
90000         1684     39  volkswagen   290
100000        2081     39  volkswagen   375
125000        4948     40  volkswagen   864
150000       30986     40  volkswagen  7113

In [50]:
autos.groupby("brand")["odometer_km"].describe().sort_index().round(1)

count      mean      std  ...       50%       75%       max
brand                                       ...                              
alfa_romeo        316.0  132199.4  33766.1  ...  150000.0  150000.0  150000.0
audi             4148.0  129282.8  39030.3  ...  150000.0  150000.0  150000.0
bmw              5278.0  132532.2  35166.4  ...  150000.0  150000.0  150000.0
chevrolet         265.0   98754.7  46865.3  ...  100000.0  150000.0  150000.0
chrysler          174.0  133850.6  34596.8  ...  150000.0  150000.0  150000.0
citroen           667.0  119445.3  40233.8  ...  150000.0  150000.0  150000.0
dacia             123.0   84268.3  48085.0  ...   80000.0  125000.0  150000.0
daewoo             72.0  122430.6  37594.9  ...  150000.0  150000.0  150000.0
daihatsu          123.0  115284.6  43637.4  ...  125000.0  150000.0  150000.0
fiat             1235.0  116825.9  42907.9  ...  125000.0  150000.0  150000.0
ford             3310.0  124758.3  39626.2  ...  150000.0  150000.0  150000.0
honda             377.0  122851.5  40245.6  ...  150000.0  150000.0  150000.0
hyundai           473.0  106511.6  45213.6  ...  125000.0  150000.0  150000.0
jaguar             74.0  121081.1  43212.4  ...  150000.0  150000.0  150000.0
jeep              108.0  127546.3  38375.3  ...  150000.0  150000.0  150000.0
kia               341.0  112434.0  44911.0  ...  125000.0  150000.0  150000.0
lada               29.0   85517.2  38250.7  ...   80000.0  100000.0  150000.0
lancia             52.0  122019.2  39373.3  ...  150000.0  150000.0  150000.0
land_rover         96.0  118489.6  46843.1  ...  150000.0  150000.0  150000.0
mazda             727.0  124745.5  39862.5  ...  150000.0  150000.0  150000.0
mercedes_benz    4556.0  131092.0  36179.2  ...  150000.0  150000.0  150000.0
mini              415.0   88602.4  43910.3  ...   90000.0  125000.0  150000.0
mitsubishi        391.0  126930.9  38808.1  ...  150000.0  150000.0  150000.0
nissan            725.0  118572.4  41694.1  ...  150000.0  150000.0  150000.0
opel             5185.0  129309.5  37379.9  ...  150000.0  150000.0  150000.0
peugeot          1417.0  127191.2  36182.3  ...  150000.0  150000.0  150000.0
porsche           284.0   98274.6  49336.8  ...  100000.0  150000.0  150000.0
renault          2271.0  128320.1  37073.7  ...  150000.0  150000.0  150000.0
rover              65.0  135615.4  30904.4  ...  150000.0  150000.0  150000.0
saab               77.0  144415.6  22197.9  ...  150000.0  150000.0  150000.0
seat              872.0  121697.2  41139.6  ...  150000.0  150000.0  150000.0
skoda             768.0  111061.2  45710.5  ...  125000.0  150000.0  150000.0
smart             668.0   99595.8  42200.4  ...  100000.0  150000.0  150000.0
sonstige_autos    475.0   89536.8  57277.3  ...  100000.0  150000.0  150000.0
subaru            105.0  124857.1  40158.7  ...  150000.0  150000.0  150000.0
suzuki            284.0  109049.3  42774.5  ...  125000.0  150000.0  150000.0
toyota            599.0  115709.5  42001.0  ...  125000.0  150000.0  150000.0
trabant            64.0   58828.1  41035.6  ...   60000.0   80000.0  150000.0
volkswagen      10172.0  128764.3  38742.7  ...  150000.0  150000.0  150000.0
volvo             437.0  139073.2  28474.2  ...  150000.0  150000.0  150000.0

[40 rows x 8 columns]

In [51]:
autos.groupby("odometer_km")["price"].describe().sort_index().round(1)

count     mean       std  ...      50%      75%         max
odometer_km                              ...                              
5000           872.0  11932.3  133842.4  ...   1100.0   5050.0   3890000.0
10000          244.0  20104.8   21125.3  ...  15925.0  27005.0    163991.0
20000          754.0  17868.7   23134.0  ...  13900.0  23767.5    345000.0
30000          763.0  16490.8   16903.9  ...  12500.0  21900.0    220000.0
40000          795.0  49738.8  968569.8  ...  12290.0  19995.0  27322222.0
50000          997.0  25922.0  353989.3  ...  10000.0  17900.0  11111111.0
60000         1122.0  12231.1   12869.9  ...   9000.0  15999.0    190000.0
70000         1186.0  10756.0    9248.8  ...   8000.0  14999.0     63999.0
80000         1386.0   9562.8    8870.1  ...   7270.0  13000.0    105000.0
90000         1684.0   8273.3    8248.7  ...   5925.0  11200.0     69997.0
100000        2081.0   7848.8   10504.4  ...   5350.0  10500.0    299000.0
125000        4948.0   6280.6   15803.5  ...   3950.0   8150.0    999999.0
150000       30986.0   7728.8  575959.7  ...   2000.0   4600.0  99999999.0

[13 rows x 8 columns]

Brand Observations Final

-Mercedes is the most expensive brand, C_klasse top model

-Top 3 brands are Mercedes, audi, and bmw

-Due to Volkswagen having a price that is an avg of the top 10, this 
might be a reason why it is the most sold brand. Top sold model is polo

-The least expensive in the top 10 are fiat and renault

-We can see that the higher milage, the less the price

In [52]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           6508
bmw                  8338
opel                 5254
mercedes_benz       30330
audi                 9088
ford                 7073
renault              2390
peugeot              3033
fiat                 2680
seat                 4301

In [53]:

brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [54]:

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132532
mercedes_benz        131091
opel                 129309
audi                 129282
volkswagen           128764
renault              128320
peugeot              127191
ford                 124758
seat                 121697
fiat                 116825

In [55]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132532        8338
mercedes_benz        131091       30330
opel                 129309        5254
audi                 129282        9088
volkswagen           128764        6508
renault              128320        2390
peugeot              127191        3033
ford                 124758        7073
seat                 121697        4301
fiat                 116825        2680